### PR TITLE
[Lock] Automaticaly release lock when user forget it

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

The objective of this PR is to automatically release the Lock when the user forget to call the `release` method (like the `fclose` function does in, php core).
The implementation is comparable to #22115 but the purpose is not the same.

I create a new PR to not mix things.

### CODE REVIEW

**Comment by stof** [comment]
this would force to have extra accesses to the store for everyone writing correct code releasing locks.
and it does not ensure it will be released: https://github.com/symfony/symfony/pull/22115#issuecomment-288594301

So I'm rather -1 on this.

the second commit adding exception messages is good though (please submit it separately)

**Comment by nicolas-grekas** [comment]
@stof should the store release the lock then?
IMHO, it'd be really nice to release any remaining lock once not used anymore.
Not as a feature one could rely on, but as a "be-clean" strategy.

**Comment by stof** [comment]
What we could do to avoid performance overhead on the store access is to add a private property storing whether we already asked to release the lock, to skip the logic in the destructor in this case (and reset this private property if we ask to acquire the lock again)

**Comment by jderusse** [comment]
I add a `isDirty` parameter to address the suggestion of @stof .
I'm not a big fan of the name of this variable, if you've a better idea, I'm listening (I don't wont to use a `isLocked` variable to avoid confusion, with the only source of trust: le Store)

**Comment by nicolas-grekas** [comment]
Note that on a command tunnel, it may be nice to allow one to "lock" some item in their "stock" accros several pages, so one should be able to opt-out from this auto-unlocking.

**Comment by jderusse** [comment]
I wonder if we should take this use-case into account @nicolas-grekas
Most of stores won't be compatible (stores linked to the process like flock, semaphore, etc..)

**Comment by fabpot** [comment]
We need to take a decision on this one. Either merge or close, but before 3.3. As first beta is probably coming out next week, what do you think? @symfony/deciders

**Comment by nicolas-grekas** [comment]
I'm personnaly in favor of having auto-unlock by default, and manual unlock as opt-in. Less space for mistakes.

**Comment by jderusse** [comment]
What do you think about
```
class LockFactory
{
  public function createLock(): AutoRealesedLock
  public function createManualLock(): Lock
}
```
Can we change the interface contract or is it considered as BC break?

**Comment by nicolas-grekas** [comment]
Do we need a new Lock type of can we deal with a new constructor/factory arg ($autoUnlock?)
About the interface, we can change everything, this has never been tagged so we are not bound by any BC promise.

**Comment by robfrawley** [comment]
I personally think it is important to be able to determine the type of lock, whether that is allowed by having separate `AutoRealesedLock` and `Lock`

```php
if (!$lock instanceof AutoRealesedLock) {
  // do stuff
}
```

or through a method on the lock instance itself

```php
if (!$lock->isAutoReleased()) {
  // do stuff
}
```

I don't have any particular preference.

**Comment by nicolas-grekas** [comment]
do we really have a use case were we would have conditional branching base on this?

**Comment by robfrawley** [comment]
@nicolas-grekas In you challenging me to provide a real-world use case for such a conditional, you convinced me that maybe there isn't a need, as I can't think of any. Carry on. ;-)

**Comment by fabpot** [comment]
What's the status of this PR? I'd like to merge/close all PRs about the lock component ASAP to avoid having to revert the component for 3.4.

**Comment by jderusse** [comment]
@fabpot All discussions are resolved. This PR is waiting for approval.

- auto release avoid unnecessary calls on stores as mentionned by @stof  
- auto release is configurable as per @nicolas-grekas's suggestion

File: src/Symfony/Component/Lock/Factory.php
Comment (fabpot) [inline src/Symfony/Component/Lock/Factory.php]: `Whether to automatically release the lock or not`

File: src/Symfony/Component/Lock/Lock.php
Comment (fabpot) [inline src/Symfony/Component/Lock/Lock.php]: same here

File: src/Symfony/Component/Lock/Lock.php
Comment (chalasr) [inline src/Symfony/Component/Lock/Lock.php]: `releases`

Comment (jderusse) [inline src/Symfony/Component/Lock/Lock.php]: fixed

File: src/Symfony/Component/Lock/Factory.php
Comment (jderusse) [inline src/Symfony/Component/Lock/Factory.php]: fixed.
Same in Lock.php

File: src/Symfony/Component/Lock/Lock.php
Comment (NoiseByNorthwest) [inline src/Symfony/Component/Lock/Lock.php]: @jderusse I think what you are tracking here is simply the ownership status of the current Lock object, which IMO should be renamed to LockHandler or something else less ambiguous with the lock which is the shared resource between many lock handlers.
So it would be great to rename "dirty" property to "owner" and add ownership concept to LockHandler API.

It will allows to:
 - ensure that an owning lock handler can only release a lock (otherwise throws LogicException).
 - ensure that a non-owning lock handler can only acquire a lock (otherwise throws LogicException).
 - expose to the user an isOwner() method (!= isAcquired())

By the way this will also help to fix an issue I've found in [CombinedStore implementation](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/CombinedStore.php#L88) where you potentially release (as a cleaning step) acquired but not owned locks.

Comment (jderusse) [reply src/Symfony/Component/Lock/Lock.php]: By design, the state of the lock is stored in the `Key` object. The `Lock` object is only a wraper and helper between the store and the key.
The following code works

```
$key = new Key(uniqid(__METHOD__, true));
$lock1 = new Lock($key, $store);
$lock2 = new Lock($key, $store);

$lock1->acquire();
$lock2->isAcquired(); // true
```

The purpose of this variable is not to track the ownership of the Lock, but to quickly know if the lock is acquired and avoid to call havy operation on the Lock destruction when the lock has not been acquired or has already been release (see comment https://github.com/symfony/symfony/pull/22132#issuecomment-289002450)

Comment (NoiseByNorthwest) [reply src/Symfony/Component/Lock/Lock.php]: My point is just that ownership must be tracked to avoid ownership issues such as (for non reentrant lock, which is the case here):
 - acquiring an already owned lock (i.e. already acquired by the current lock object).
 - releasing a non owned lock (even and especially if owned by someone else).

This kind of issue can arise as long as program flaws exist, and when it happens throwing something like a LogicException is far more better than corrupt the system IMO.

And I think there is an ownership issue [right here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/CombinedStore.php#L88).

Comment (jderusse) [reply src/Symfony/Component/Lock/Lock.php]: Unicity (or ownership) is tracked and stored in the `Key` by the store (A Key is not a resource
You can't release a Lock with a different key.

The ownership is handled:
 - by storing the filepointer in flock https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/FlockStore.php#L123
 - by storing/and comparing a random value in Memcached and Redis https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/RedisStore.php#L105

Stores won't throw exception if you try to delete a key which is not the owner

Comment (jderusse) [reply src/Symfony/Component/Lock/Lock.php]: Unicity (or ownership) is tracked and stored in the `Key` by the store (A Key is not a resource)
You can't release a Lock with a different key.

The ownership is handled:
 - by storing the filepointer in flock https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/FlockStore.php#L123
 - by storing/and comparing a random value in Memcached and Redis https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/RedisStore.php#L105

Stores won't throw exception if you try to delete a key which is not the owner, but they neither don't deleted the lock.

```
$key1 = new Key('foo');
$key2 = new Key('foo');

$lock1 = new Lock($store, key1);
$lock2 = new Lock($store, key2);

$lock1->acquire(); // return true
$lock2->acquire(); // return false

$lock2->release(); // do nothing
$lock2->isAcquire(); // return false
$lock1->isAcquire(); // return true
$lock2->acquire(); // return false
```

Comment (NoiseByNorthwest) [reply src/Symfony/Component/Lock/Lock.php]: Ok nice, but I'm just surprised to have ownership implemented on lock provider side (stores) since it could be simply and in a single place tracked in Lock class with a bool flag.

So the behavior [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/CombinedStore.php#L88) is indeed correct since non owned acquired locks will be silently discarded.
Correct but just still hard to understand IMO (I'd prefer explicitly checking isOwner() before calling release()).

And I still find very useful to have LogicException thrown when I make something illogic :smile:

File: src/Symfony/Component/Lock/Lock.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Lock.php]: is the try catch actually required?

Comment (jderusse) [reply src/Symfony/Component/Lock/Lock.php]: to not throw an exception when the auto-release try to release the lock...

leave the try/cacth and you may miss exception when you count on the auto-release to release the lock
remove the try/catch and you may have an unexpected exception when you manually release the lock and the auto-release retry to do it

Comment (nicolas-grekas) [reply src/Symfony/Component/Lock/Lock.php]: isn't "release" idempotent?

Comment (jderusse) [reply src/Symfony/Component/Lock/Lock.php]: indeed. Fixed

File: src/Symfony/Component/Lock/Factory.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Factory.php]: Whether to automatically release the lock or not when the lock instance is destroyed

File: src/Symfony/Component/Lock/Lock.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Lock.php]: when... (see previous comment)

File: src/Symfony/Component/Lock/Lock.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Lock.php]: one "if" is enough in the method:
`if ($this->autoRelease && $this->dirty && $this->isAcquired()) { //...`

File: src/Symfony/Component/Lock/Lock.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Lock.php]: unrelated, but just wondering: isn't this creating a race condition?

Comment (jderusse) [reply src/Symfony/Component/Lock/Lock.php]: No, unless you share the key with an other process (i'm talking about the key, not the resource).

The method `exists` check both the "resource" and the "owner". You may have a race condition by using pcntl_fork (or serializeing then share the key) and both processes try to release and acquire the lock on the same time. But IMO, that's a weird case.

BTW, we don't provide methods to check if the lock is acquired by someone else.

**Comment by fabpot** [comment]
What's the status of this PR? Mergeable? That's probably a PR we need to close or merge before feature freeze.

**Comment by jderusse** [comment]
All comments are take into account. IMO ready to merge.ping @symfony/deciders

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jderusse.

**Pull request closed without merging**
